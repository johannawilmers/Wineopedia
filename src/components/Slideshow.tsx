import React from "react";
import { Wine } from "../utils/types";
import { useFavorites } from "../hooks/useFavorites";
import { useFilter } from "../hooks/useFilter";
import "../styles/slideshow.css";
import "../styles/listView.css";

interface WineSlideShow {
  wines: Wine[];
  currentIndex: number;
  activeButton: string;
  setCurrentIndex: (index: number) => void;
}

/**
 * This component is used to display a slideshow of wines, showing one wine at a time
 * @param {WineSlideShow} wines
 */
const Slideshow: React.FunctionComponent<WineSlideShow> = ({ wines }) => {
  const { favorites, toggleFavorite } = useFavorites();
  const { currentIndex, setCurrentIndex } = useFilter();
  if (wines.length === 0) {
    return (
      <div id="errorNoWine">
        <p>No products in this category. Add your favorites in homepage</p>
      </div>
    );
  }
  const currentWine = wines[currentIndex]?.basic.productId
    ? wines[currentIndex] || {}
    : wines[0];
  const isFavorite =
    currentWine.basic && favorites.includes(currentWine.basic.productId);

  const getPreviousElement = () => {
    currentIndex !== 0 ? setCurrentIndex(currentIndex - 1) : setCurrentIndex(0);
  };

  const getNextElement = () => {
    currentIndex < wines.length - 1
      ? setCurrentIndex(currentIndex + 1)
      : setCurrentIndex(0);
  };

  return (
    <div>
      <div className="slideShow">
        <button
          className="buttonLeft"
          aria-label="button-ClickLeft"
          onClick={getPreviousElement}
        />
        <div className="wineSlide">
          <h2 className="slideshowProductName">
            {currentWine.basic?.productShortName ||
              "Product Name Not Available"}
          </h2>
          {currentWine.basic && (
            <img
              className="wineImage"
              src={`https://bilder.vinmonopolet.no/cache/300x300-0/${currentWine.basic.productId}-1.jpg`}
              alt={currentWine.basic.productShortName}
            />
          )}
        </div>
        <button
          className="buttonRight"
          aria-label="button-ClickRight"
          onClick={getNextElement}
        />
      </div>
      <div className="wineInformation">
        <p>
          <b> ProductID: </b> {currentWine.basic?.productId || "N/A"}
        </p>
        <p>
          <b> Last Changed at: </b> {currentWine.lastChanged?.date}{" "}
          {currentWine.lastChanged?.time}
        </p>
        <div className="buttonContainer">
          <button
            onClick={() => toggleFavorite(currentWine.basic?.productId)}
            className="favoriteButton"
          >
            {isFavorite ? "Remove from Favorites" : "Add to Favorites"}
          </button>
        </div>
      </div>
    </div>
  );
};

export default Slideshow;
