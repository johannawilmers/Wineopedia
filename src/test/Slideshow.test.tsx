import { render, fireEvent, RenderResult } from "@testing-library/react";
import { BrowserRouter } from "react-router-dom";
import Slideshow from "../components/Slideshow";
import "@testing-library/jest-dom/extend-expect";
import { screen } from "@testing-library/dom";

const firstwine = {
  basic: {
    productId: "133501",
    productShortName: "Tissot Vin Jaune La Vasée",
    favorite: false,
  },
  lastChanged: {
    date: "2021-04-16",
    time: "06:19:05",
  },
};

const secondwine = {
  basic: {
    productId: "162701",
    productShortName: "Niepoort Bioma Vintage",
    favorite: false,
  },
  lastChanged: {
    date: "2021-03-15",
    time: "22:57:42",
  },
};

const thirdWine = {
  basic: {
    productId: "620302",
    productShortName: "Gaia Vin Santo",
    favorite: false,
  },
  lastChanged: {
    date: "2021-03-15",
    time: "22:57:42",
  },
};

const testArray = [firstwine, secondwine, thirdWine];

describe("renders slideshow component", () => {
  let wrapper: RenderResult;

  beforeEach(() => {
    wrapper = render(
      <BrowserRouter>
        <Slideshow
          wines={testArray}
          currentIndex={0}
          activeButton={"All wines"}
          setCurrentIndex={function (): void {
            throw new Error("Function not implemented.");
          }}
        />
      </BrowserRouter>,
    );
  });

  it("test that firstwine is shown in slideshow", () => {
    const firstWine = wrapper.getByText("Tissot Vin Jaune La Vasée");
    expect(firstWine).toBeInTheDocument();
    expect(firstWine).toMatchSnapshot();
  });

  it("Should be able to click to next and prev element in the slideshow", () => {
    const firstWine = wrapper.getByText("Tissot Vin Jaune La Vasée");
    expect(firstWine).toBeInTheDocument();

    // click to the next wine in slideshow
    const changeSlide = wrapper.getByRole("button", {
      name: /button-ClickRight/i,
    });
    fireEvent.click(changeSlide);

    const secondWine = wrapper.getByText("Niepoort Bioma Vintage");
    expect(secondWine).toBeInTheDocument();

    const theRemovedWine = screen.queryByText("Tissot Vin Jaune La Vasée");
    expect(theRemovedWine).not.toBeInTheDocument();

    //clicks back to the first wine in the slideshow
    const changeSlideBack = wrapper.getByRole("button", {
      name: /button-clickLeft/i,
    });
    fireEvent.click(changeSlideBack);
    expect(firstWine).toBeInTheDocument();
  });

  it("should be able to like/unlike a wine", () => {
    const favButton = screen.getByText("Add to Favorites");
    fireEvent.click(favButton);
    expect(favButton.textContent).toBe("Remove from Favorites");
    expect(favButton.textContent).not.toBe("Add to Favorites");
    fireEvent.click(favButton);
    expect(favButton.textContent).toBe("Add to Favorites");
  });

  it("Should show an empty slideshow and the text 'Product Name Not Available'", () => {
    wrapper.rerender(
      <BrowserRouter>
        <Slideshow
          wines={[]}
          currentIndex={0}
          activeButton={""}
          setCurrentIndex={function (): void {
            throw new Error("Function not implemented.");
          }}
        ></Slideshow>
      </BrowserRouter>,
    );
    expect(
      wrapper.getByText(
        "No products in this category. Add your favorites in homepage",
      ),
    ).toBeInTheDocument();
  });

  it("Should show an empty slideshow and the text 'Product Name Not Available'", () => {
    wrapper.rerender(
      <BrowserRouter>
        <Slideshow
          wines={[]}
          currentIndex={0}
          activeButton={""}
          setCurrentIndex={function (): void {
            throw new Error("Function not implemented.");
          }}
        ></Slideshow>
      </BrowserRouter>,
    );
    expect(
      wrapper.getByText(
        "No products in this category. Add your favorites in homepage",
      ),
    ).toBeInTheDocument();
  });
});
