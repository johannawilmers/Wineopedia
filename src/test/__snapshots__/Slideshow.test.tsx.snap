// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`renders slideshow component > test that firstwine is shown in slideshow 1`] = `
<h2
  class="slideshowProductName"
>
  Tissot Vin Jaune La Vas√©e
</h2>
`;
